packages/ui/src/views/run-pipeline/utils/inputs-utils.ts

```diff
@@ -4,19 +4,95 @@ import * as z from 'zod'
 
 import { IInputDefinition, unsetEmptyStringOutputTransformer } from '@harnessio/forms'
 
+import { type InputLayout } from './types'
+
 /** pipeline inputs to form inputs conversion */
-export function pipelineInputs2FormInputs(
-  pipelineInputs: Record<string, any>,
+export function pipelineInputs2FormInputs({
+  pipelineInputs,
+  options,
+  pipelineInputLayout = []
+}: {
+  pipelineInputs: Record<string, any>
   options: { prefix?: string }
-): IInputDefinition[] {
-  const formInputs: IInputDefinition[] = []
+  pipelineInputLayout?: InputLayout
+}): IInputDefinition[] {
+  /**
+   * Pre-process inputs for valid layout.
+   * Rule 1 - If there are duplicate keys in the layout, we skip the layout and return all inputs flat
+   * Rule 2 - ...
+   */
+  const duplicateKeys = validateUniqueInputKeysInLayout(pipelineInputLayout)
+
+  // If duplicates found, skip layout and return all inputs flat
+  if (validateUniqueInputKeysInLayout(pipelineInputLayout).length > 0) {
+    console.warn('Duplicate input keys detected in layout. Using flat input list instead. Keys:', duplicateKeys)
+
+    const fallbackInputs: IInputDefinition[] = []
+    forOwn(pipelineInputs, (value, key) => {
+      fallbackInputs.push(pipelineInput2FormInput(key, value, options))
+    })
+    return fallbackInputs
+  }
+
+  const processedInputKeys = new Set<string>()
+
+  const processLayout = (layout: InputLayout): IInputDefinition[] => {
+    return layout.flatMap(item => {
+      if (typeof item === 'string') {
+        processedInputKeys.add(item)
+        const value = pipelineInputs[item]
+        return pipelineInput2FormInput(item, value, options)
+      }
+
+      // If group has no title, flatten its items
+      if (!item.title) {
+        return processLayout(item.items)
+      }
+
+      return {
+        inputType: 'group',
+        path: '',
+        label: item.title,
+        inputs: processLayout(item.items),
+        inputConfig: {
+          autoExpandGroups: item.open
+        }
+      }
+    })
+  }
+
+  const inputsFromLayout = processLayout(pipelineInputLayout)
 
+  const remainingInputs: IInputDefinition[] = []
   forOwn(pipelineInputs, (value, key) => {
-    const formInput = pipelineInput2FormInput(key, value, options)
-    formInputs.push(formInput)
+    if (!processedInputKeys.has(key)) {
+      remainingInputs.push(pipelineInput2FormInput(key, value, options))
+    }
   })
 
-  return formInputs
+  return [...inputsFromLayout, ...remainingInputs]
+}
+
+export const validateUniqueInputKeysInLayout = (layout: InputLayout): string[] => {
+  const inputOccurrences = new Map<string, number>()
+
+  const traverse = (layout: InputLayout) => {
```

Comment:
can this be a pure function outside of `validateUniqueInputKeysInLayout` as well?

Reply:
I agree this feels cleaner and allows reuse, but I'm curious what your reason for this suggestion is. Same as I listed or something else?

Reply:
Also done.